### ISSUE TITLE
Error: Add search attribute failed

### ISSUE THREAD

**Issue by pauldemarco** [temporalio/temporal#505]
## Expected Behavior
I should be able to add a new search attribute "UserID" of type Keyword to Temporal for filtering.

## Actual Behavior
I initially receive an error when trying to add attribute. Upon listing added keys, I see the newly added key but it is of the wrong type (String).

## Steps to Reproduce the Problem

  1. Deploy the latest Temporal 0.26.0 with Elasticsearch enabled.
  2. Add a new search attribute by running the following command: `tctl --namespace default adm cl asa --search_attr_key UserID --search_attr_type 1`
  3. Observe the following error and stack trace:
```
bash-5.0# tctl  --namespace default adm cl asa --search_attr_key UserID --search_attr_type 1
Are you trying to add key [UserID] with Type [Keyword]? Y/N y

Error: Add search attribute failed.
Error Details: Failed to update dynamic config, err: unable to update key.
Stack trace:
goroutine 1 [running]:
runtime/debug.Stack(0xd, 0x0, 0x0)
	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
runtime/debug.PrintStack()
	/usr/local/go/src/runtime/debug/stack.go:16 +0x22
github.com/temporalio/temporal/tools/cli.printError(0x1c5008f, 0x1c, 0x2003cc0, 0xc00000c300)
	/temporal/tools/cli/util.go:526 +0x2ad
github.com/temporalio/temporal/tools/cli.ErrorAndExit(0x1c5008f, 0x1c, 0x2003cc0, 0xc00000c300)
	/temporal/tools/cli/util.go:537 +0x49
github.com/temporalio/temporal/tools/cli.AdminAddSearchAttribute(0xc000127b80)
	/temporal/tools/cli/adminClusterCommands.go:61 +0x50f
github.com/temporalio/temporal/tools/cli.newAdminClusterCommands.func1(0xc000127b80)
	/temporal/tools/cli/admin.go:812 +0x2b
github.com/urfave/cli.HandleAction(0x18cf500, 0x1cc9b88, 0xc000127b80, 0xc000127b80, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/app.go:528 +0x7c
github.com/urfave/cli.Command.Run(0x1c3692d, 0xf, 0x0, 0x0, 0xc0006998c0, 0x1, 0x1, 0x1c4d6ff, 0x1a, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/command.go:174 +0x57a
github.com/urfave/cli.(*App).RunAsSubcommand(0xc0004d9c00, 0xc000127760, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/app.go:407 +0x915
github.com/urfave/cli.Command.startApp(0x1c2a872, 0x7, 0x0, 0x0, 0xc0006999f0, 0x1, 0x1, 0x1c551de, 0x1e, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/command.go:373 +0x845
github.com/urfave/cli.Command.Run(0x1c2a872, 0x7, 0x0, 0x0, 0xc0006999f0, 0x1, 0x1, 0x1c551de, 0x1e, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/command.go:102 +0xa2b
github.com/urfave/cli.(*App).RunAsSubcommand(0xc0004d9a40, 0xc000127600, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/app.go:407 +0x915
github.com/urfave/cli.Command.startApp(0x1c27ae6, 0x5, 0x0, 0x0, 0xc000699970, 0x1, 0x1, 0x1c3df67, 0x13, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/command.go:373 +0x845
github.com/urfave/cli.Command.Run(0x1c27ae6, 0x5, 0x0, 0x0, 0xc000699970, 0x1, 0x1, 0x1c3df67, 0x13, 0x0, ...)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/command.go:102 +0xa2b
github.com/urfave/cli.(*App).Run(0xc0004d96c0, 0xc00010c000, 0xa, 0xa, 0x0, 0x0)
	/go/pkg/mod/github.com/urfave/cli@v1.22.4/app.go:279 +0x7c7
main.main()
	/temporal/cmd/tools/cli/main.go:37 +0x4b
```
4. List the search attributes and notice the wrong type:
```
bash-5.0#  tctl --namespace default cl get-search-attr
+-----------------------+------------+
|          KEY          | VALUE TYPE |
+-----------------------+------------+
| BinaryChecksums       | Keyword    |
| CloseTime             | Int        |
| CustomBoolField       | Bool       |
| CustomDatetimeField   | Datetime   |
| CustomDoubleField     | Double     |
| CustomIntField        | Int        |
| CustomKeywordField    | Keyword    |
| CustomStringField     | String     |
| ExecutionStatus       | Int        |
| ExecutionTime         | Int        |
| HistoryLength         | Int        |
| NamespaceId           | Keyword    |
| RunId                 | Keyword    |
| StartTime             | Int        |
| TaskQueue             | Keyword    |
| TemporalChangeVersion | Keyword    |
| UserID                | String     |
| WorkflowId            | Keyword    |
| WorkflowType          | Keyword    |
+-----------------------+------------+
```
## Additional Questions
Now that I have an incorrect attribute added, is there any way to edit or remove it?
Does the search attribute need to be registered explicitly by the CLI before it can be used elsewhere (go-sdk)?  The docs mention something here but it's still a bit unclear: https://docs.temporal.io/docs/learn-workflow-filtering#search-attributes-go-client-usage

## Specifications

  - Version: 0.26.0
  - Platform: Ubuntu 18.04

**Comment by pauldemarco** [comment]
I've just noticed that the attribute type registered is always 1 less than specified. For example, if I specify type 1 (Keyword), it registers as type 0 (String). Perhaps some indexing issue with the list?

It should be [0:String, 1:Keyword, 2:Int, 3:Double, 4:Bool, 5:Datetime] (default: -1)

**Comment by pauldemarco** [comment]
It seems I cannot use the search attribute after all. I see the following failure when I attempt to upsert search attributes from go code:
![image](https://user-images.githubusercontent.com/16725750/86959172-590cff80-c12b-11ea-9068-015ade8e7e89.png)

**Comment by alexshtin** [comment]
As for indexes, doc is wrong. 0 is now UNSPECIFIED and the should start with one. We need to update doc.

**Comment by samarabbas** [comment]
@cullywakelin can you look into updating the docs to reflect this?

**Comment by pauldemarco** [comment]
Would anyone be able to verify they are also seeing the error from Step 3?

**Comment by mastermanu** [comment]
Hey @pauldemarco, we were not able to reproduce the error you got, at least against a docker container deployed on the local environment. Do you have temporal-frontend logs you can share?

**Comment by pauldemarco** [comment]
@mastermanu These are the logs created in temporal-frontend after running `tctl --namespace default adm cl asa --search_attr_key UserID --search_attr_type 1`
```
{"level":"warn","ts":"2020-07-08T21:39:12.532Z","msg":"error in prometheus reporter","error":"a previously registered descriptor with the same fully-qualified name as Desc{fqName: \"service_latency\", help: \"service_latency histogram\", constLabels: {}, variableLabels: [type operation]} has different label names or a different help string","logging-call-at":"metrics.go:135"}
{"level":"warn","ts":"2020-07-08T21:39:12.532Z","msg":"error in prometheus reporter","error":"a previously registered descriptor with the same fully-qualified name as Desc{fqName: \"service_requests\", help: \"service_requests counter\", constLabels: {}, variableLabels: [type operation]} has different label names or a different help string","logging-call-at":"metrics.go:135"}
{"level":"error","ts":"2020-07-08T21:39:12.532Z","msg":"Internal service error","service":"frontend","error":"Failed to update dynamic config, err: unable to update key.","logging-call-at":"adminHandler.go:1170","stacktrace":"github.com/temporalio/temporal/common/log/loggerimpl.(*loggerImpl).Error\n\t/temporal/common/log/loggerimpl/logger.go:138\ngithub.com/temporalio/temporal/service/frontend.(*AdminHandler).error\n\t/temporal/service/frontend/adminHandler.go:1170\ngithub.com/temporalio/temporal/service/frontend.(*AdminHandler).AddSearchAttribute\n\t/temporal/service/frontend/adminHandler.go:165\ngithub.com/temporalio/temporal/service/frontend.(*AdminNilCheckHandler).AddSearchAttribute\n\t/temporal/service/frontend/adminNilCheckHandler.go:113\ngithub.com/temporalio/temporal/.gen/proto/adminservice/v1._AdminService_AddSearchAttribute_Handler.func1\n\t/temporal/.gen/proto/adminservice/v1/service.pb.go:579\ngithub.com/temporalio/temporal/service/frontend.interceptor\n\t/temporal/service/frontend/service.go:323\ngithub.com/temporalio/temporal/.gen/proto/adminservice/v1._AdminService_AddSearchAttribute_Handler\n\t/temporal/.gen/proto/adminservice/v1/service.pb.go:581\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834"}
```

**Comment by mastermanu** [comment]
Ok, let me dig into this and get back to you regarding the error you are getting. 

On a separate FYI, we're landing a change so that the --search_attr_type parameter takes in the actual case-insensitive type-name (e.g. string, keyword, int, etc) so that you don't have to look at a reverse mapping to an integer. We were able to confirm that adding attributes / searching workflows by those attributes worked properly with that change.

**Comment by mastermanu** [comment]
So an update here:

1) We assume that you are using our helm charts to deploy Temporal onto a Kubernetes cluster. If this is not the case, let us know, but the rest of the explanation should still be correct.

2) The yaml element called "dynamicConfigClient" within the configuration template (and therefore the configuration file) is not present. (See /etc/temporal/config/docker.yaml on your running node). Because that element is absent, the dynamic config client we generate is the "no-op" client (as you discovered above). A no-op client rejects any attempts to change it.

3) For a deployed cluster with more than one temporal server node, even if we specified the dynamicConfigClient configuration, if you were to run the tctl command to create a new search attribute key/value type, it would persist the change to ONLY the node which that tctl command hit...but not any of the other nodes. Also if that pod were to be recycled / re-imaged, that change would be lost.

4) Thus, we are currently at a point where the included dynamicconfiguration file has to be set in the docker container image itself or copied over on startup.

In other words, you would create the dynamicconfiguration file ahead of time (with whatever additional attributes you want indexed) and then make sure it is included in the temporal server container image you are deploying. If you ssh into one of your server nodes, you'll notice that the /etc/temporal/config/dynamicconfig/development_es.yaml already exists. So you'd be adding the appropriate lines to that file, add it to your temporal server image, and then deploy that image.

You also need to make sure the dynamicConfigClient yaml element is added to the config-template file.
See here: https://github.com/temporalio/temporal-helm-charts/blob/master/templates/server-configmap.yaml (it would be added just below the publicClient element, at the same level of indentation as publicClient with the following fields, like so):

publicClient:
   ...

dynamicConfigClient:
    filepath: config/dynamicconfig/development_es.yaml
    pollInterval: "60s"

You can also run the tctl tool on a local dev box environment to add all the attributes, then grab the dynamic config file from the local container running locally and then copy it over if you don't want to edit the file by hand. For instance, this is what the file looked like when I added UserID as a key and Keyword as the Type:

frontend.enableClientVersionCheck:
- value: true
  constraints: {}
frontend.validSearchAttributes:
- value:
    BinaryChecksums: Keyword
    CloseTime: Int
    CustomBoolField: Bool
    CustomDatetimeField: Datetime
    CustomDoubleField: Double
    CustomIntField: Int
    CustomKeywordField: Keyword
    CustomNamespace: Keyword
    CustomStringField: String
    ExecutionTime: Int
    HistoryLength: Int
    NamespaceId: Keyword
    Operator: Keyword
    RolloutId: Keyword
    RunId: Keyword
    StartTime: Int
    Status: Int
    TaskQueue: Keyword
    TemporalChangeVersion: Keyword
    **UserID: 2**
    WorkflowId: Keyword
    WorkflowType: Keyword
    addon: Keyword
    addon-type: Keyword
    environment: Keyword
    project: Keyword
    service: Keyword
    user: Keyword
  constraints: {}
system.advancedVisibilityWritingMode:
- value: "on"
  constraints: {}
system.enableReadVisibilityFromES:
- value: true
  constraints: {}
system.minRetentionDays:
- value: 0
  constraints: {}

It is on our roadmap to:
    1) Enable dynamic configuration of a multi-node cluster via the tctl tool (e.g. enabling you to add search attributes on the fly)
    2) To support removing search attribute keys (which was another question you also asked).

Let us know if we can provide any further clarifications here or provide other assistance.

**Comment by pauldemarco** [comment]
@mastermanu  Correct, I am using the official Temporal helm chart and deploying to a dev server running kubernetes.


I've added the dynamicConfigClient to the configmap and re-deployed: https://github.com/pauldemarco/temporal-helm-charts/commit/bae60f73a445c19aa5791b6221893042369c7af7

I verified it shows up by running the following on temporal-frontend pod:
`bash-5.0# cat /etc/temporal/config/docker.yaml`

I can now also successfully register a new type on the admin-tools pod:
```
bash-5.0# tctl --namespace default adm cl asa --search_attr_key UserID --search_attr_type 2
Are you trying to add key [UserID] with Type [Int]? Y/N y

Success
```

I am still however receiving errors when running the workflow:
![image](https://user-images.githubusercontent.com/16725750/86995390-e32e8580-c176-11ea-8d2f-6dc4e8b61898.png)

Reading out the file `bash-5.0# cat /etc/temporal/config/dynamicconfig/development_es.yaml ` shows the new key added only for the pod where `tctl` was run, like you mentioned.

At this point, it seems I need to include this custom **development_es.yaml** file with startup. I haven't yet had to customize the temporal server image that I use with the helm chart, and I'm not exactly sure how. I usually just increment the version number (currently 0.26.0). Any suggestions?

**Comment by cullywakelin** [comment]
The docs have been updated to reflect @mastermanu's adding a new **search-attr-type** change via CLI: http://docs.temporal.io/docs/learn-workflow-filtering/#allow-listing-search-attributes

**Comment by mastermanu** [comment]
Hey @pauldemarco,

You would do the following:
 1) Run temporal on your local dev box using docker-compose (follow instructions here: https://docs.temporal.io/docs/installing-server)
 2) Either use docker to or copy an updated deployment_es into the temporal server image, or exec bash INTO the running container and update deployment_es.yaml as you want to.
 3) Commit / tag that modified server image locally (https://www.scalyr.com/blog/create-docker-image/)
 4) Upload the image to your repository dockerhub (https://ropenscilabs.github.io/r-docker-tutorial/04-Dockerhub.html)
 5) Change the helm values.yaml file (https://github.com/temporalio/helm-charts/blob/master/values.yaml) under the "server/image" section to point to your repository in DockerHub (or you can override using the --set command line option on helm install command)

Deploy the helm chart.

For the heck of it, I went through the process and created an updated version of 0.26.0 that adds in a "UserID" attribute with a a "Keyword" type here: https://hub.docker.com/repository/docker/mastermanu2004/server

So the helm chart would look like this then:

```
server:
    enabled: true
    image:
        repository: mastermanu2004/server
        tag: latest
        pullPolicy: IfNotPresent
```

I would not rely on that image itself (although you can try doing a test deployment with it), but that should give you an idea of how to proceed.

Let us know if you have any other questions.

**Comment by mastermanu** [comment]
FYI, we will try and make the injection of dynamic configuration more seamless as well so that you don't have to create a custom image as has been outlined above. For now, the above steps should unblock you for experimentation purposes though.

**Comment by pauldemarco** [comment]
Thanks for the writeup. I had to pull down **docker-compose-es.yml** instead of the default in step 1. I created the image here after registering my custom keys: https://hub.docker.com/layers/pauldemarco/paul_temporal/paul_temporal/images/sha256-398e7f7e36fb67fee888b3c50d8bb63d049c2014bc679b8973d859ff16b12dcb?context=repo

I've pointed the helm chart to my custom image and deployed to kubernetes.

All services come up successfully. I can shell into temporal-frontend and verify the development_es.yaml contains my custom keys.

Temporal-web is now showing this error message:
![image](https://user-images.githubusercontent.com/16725750/87080779-41e71400-c1f6-11ea-8515-b7c66dc8d6c2.png)

Checking the logs for temporal-frontend show the following spam:
```
+ cqlsh --cqlversion=3.4.4 cassandra
Connection error: ('Unable to connect to any servers', {'92.242.140.21:9042': error(None, "Tried connecting to [('92.242.140.21', 9042)]. Last error: timed out")})
+ echo 'waiting for cassandra to start up'
waiting for cassandra to start up
+ sleep 1
+ cqlsh --cqlversion=3.4.4 cassandra
Connection error: ('Unable to connect to any servers', {'92.242.140.21:9042': error(None, "Tried connecting to [('92.242.140.21', 9042)]. Last error: timed out")})
+ echo 'waiting for cassandra to start up'
+ sleep 1
waiting for cassandra to start up
+ cqlsh --cqlversion=3.4.4 cassandra
Connection error: ('Unable to connect to any servers', {'92.242.140.21:9042': error(None, "Tried connecting to [('92.242.140.21', 9042)]. Last error: timed out")})
+ echo 'waiting for cassandra to start up'
+ sleep 1
waiting for cassandra to start up
+ cqlsh --cqlversion=3.4.4 cassandra
```

Is it possible when I created the docker image it somehow "snapshotted" incorrect connection info for cassandra?

**Comment by mastermanu** [comment]
Can you paste the startup logs for the temporal-frontend along with the config_template.yaml and docker.yaml file that is in that pod's configuration directory?

Also, I assume if you switch back to a "regular" image, then this problem goes away? We can compare the docker.yaml file from a functional deployment and a non-functional one to see what the difference might be here.

@markmark206 might also be able to provide better insight here as he is a bigger expert in this area :)

**Comment by pauldemarco** [comment]
> Can you paste the startup logs for the temporal-frontend along with the config_template.yaml and docker.yaml file that is in that pod's configuration directory?

If the below guide is way off, or you feel it would still be constructive for me to supply this info, let me know.

> Also, I assume if you switch back to a "regular" image, then this problem goes away? We can compare the docker.yaml file from a functional deployment and a non-functional one to see what the difference might be here.

Correct.

### Setting up custom search attributes

These are the steps that I've taken to get everything up and running:

1. Clone the temporal repo
```
git clone git@github.com/temporalio/temporal
git checkout v0.26.0
```

2. Add the custom keys to config/dynamicconfig/development_es.yaml
```
frontend.enableClientVersionCheck:
  - value: true
system.advancedVisibilityWritingMode:
  - value: "on"
system.enableReadVisibilityFromES:
  - value: true
frontend.validSearchAttributes:
  - value:
      NamespaceId: "Keyword"
      WorkflowId: "Keyword"
      RunId: "Keyword"
      WorkflowType: "Keyword"
      StartTime: "Int"
      ExecutionTime: "Int"
      CloseTime: "Int"
      Status: "Int"
      HistoryLength: "Int"
      TaskQueue: "Keyword"
      CustomStringField: "String"
      CustomKeywordField: "Keyword"
      CustomIntField: "Int"
      CustomDoubleField: "Double"
      CustomBoolField: "Bool"
      CustomDatetimeField: "Datetime"
      project: "Keyword"
      service: "Keyword"
      environment: "Keyword"
      addon: "Keyword"
      addon-type: "Keyword"
      user: "Keyword"
      CustomNamespace: "Keyword"
      Operator: "Keyword"
      RolloutId: "Keyword"
      TemporalChangeVersion: "Keyword"
      BinaryChecksums: "Keyword"
      UserID: "Keyword"
system.minRetentionDays:
    - value: 0
```

3. Build the docker images
```
make docker-images
```

4. Tag and push to docker hub
```
docker images
docker tag 3a6eef54b87a pauldemarco/paul_temporal:0.26.0-config2
docker push pauldemarco/paul_temporal
```

5. Update the helm chart to point to this new image (values.yaml)
```
server:
  enabled: true
  image:
    repository: pauldemarco/paul_temporal
    tag: 0.26.0-config2
    pullPolicy: IfNotPresent
```

6. Redeploy the helm chart
```
helm -n temporal install temporal . --timeout 900s
```

### Issue: custom keys returns zero results

At this point everything is starting properly and I'm able to successfully upsert attributes from workflows:
![image](https://user-images.githubusercontent.com/16725750/87096473-ed04c700-c210-11ea-8dde-9a938f8b2c65.png)

Unfortunately, performing any queries for my custom keys returns zero results. I've even set CustomKeywordField and UserID to the same value in my go code:
```
attr := map[string]interface{}{
		"UserID":             "user1234",
		"CustomKeywordField": "user1234",
	}
```
`tctl --ns default wf list -q 'CustomKeywordField="user1234"'`
Returns results.
`tctl --ns default wf list -q 'UserID="user1234"'`
Returns zero results.

I'm stumped again, sorry!

**Comment by mastermanu** [comment]
Hey Paul,

First of all, the way of building the docker image is much preferred over that short-cut I mentioned earlier. Thanks for pointing this out...I will update my answer to incorporate your suggestion to just clone the repo and build directly.

Secondly, I forgot one more vital step: In addition to updating the development_es.yaml file, you also have to update the schema/elasticsearch/visibility/index_template.json file so that it actually adds the index to the Elastic Search cluster when the Temporal server boots up. I went ahead and added a key/value pair "UserID" : { "type": {"keyword"} to the properties dictionary under "Attr" and tested locally, and after that, I was able to query workflows using the property.

So TL;DR, it looks like there are TWO files that have to be customized on the server image:

development_es.yaml
schema/elasticsearch/visibility/index_template.json

Really sorry for the inconvenience here. That should fix it, and we'll look to make this clearer via documentation and have a more seamless way to configure these types of things in the future.

**Comment by pauldemarco** [comment]
That did the trick!

> have a more seamless way to configure these types of things in the future.

Yes, although currently fine for experimentation, having to recompile the image to add a elasticsearch attribute is a bit of a pain, especially during development where these are constantly changing.  It would be great if these could be changed either on the fly through tctl, or getting some first class Kubernetes support through a configmap or persistent volume.

Thanks a bunch for the help Manu!

**Comment by johanforssell** [comment]
If you happen to find this issue today, this is no longer relevant (might be obvious, I don't know).
As per https://github.com/temporalio/temporal/issues/1442#issuecomment-878705053 and https://github.com/temporalio/temporal/issues/1442#issuecomment-878705053, you have to use tctl now.
